Bubble chart

panda01 opened the pull request:
Made the Line Chart convertible into a bubble chart.

Added a z aspect to the point series.
Modified the tooltip that shows up on hover to show multiple metrics.
Added adjustable Point radius to the LineChart.
Added a flag that turns off and on the lines connecting points.
Added a an adjuster for the Ratio of the smallest circle to the largest.
Modified the editor aggregations to respect the minimum attr on Agg Schemas.
Modified some tests to accommodate aforementioned changes.

Closes #2914.

lukasolson commented:
I'm not really sure how this *should* function, but I'm seeing some strange behavior when you start messing around with the number of y-axis aggregations.

If you add a dot size aggregation, then remove the y-axis aggregation, it doesn't seem to have an effect. The y-axis aggregation still shows up as if you hadn't removed it. (Again, I'm not sure was I'd expect to see with a dot size aggregation with no y-axis aggregation.)

![no-y-axis](https://cloud.githubusercontent.com/assets/1178348/6399814/fed13680-bdb1-11e4-9a04-635343297233.gif)

If you have multiple y-axis aggregations, and then add a dot size aggregation, it seems as though the dot sizes no longer are affected. It's as if you had no dot size aggregation at all. (I don't know what to do in this situation either... should the dot sizes vary on both of the y-axis lines?...)

![two-y-axes](https://cloud.githubusercontent.com/assets/1178348/6399831/26bd19fc-bdb2-11e4-9344-6efe25a2a18c.gif)

Might be worth asking some of the other guys. @jthomassie? @stormpython? @rashidkpc?

rashidkpc commented:
We should also get the result of the dot size agg into the tooltip: 
![image](https://cloud.githubusercontent.com/assets/1250387/6422976/551a00fe-be9c-11e4-831f-fc30985f9661.png)

simianhacker commented:
### Aggregations

We should remove `standard deviation` and `percentiles` metrics aggs from  the dot metric. It doesn't make sense since the dots don't support multiple metric aggs.

### Bubble Size

Is there a way to increase the max and min bubble size? Or make it configurable?

![image](https://cloud.githubusercontent.com/assets/41702/6454938/3c3bcab0-c10b-11e4-87e1-04e85a59f87f.png)

### Much Colors, So Pretty

![image](https://cloud.githubusercontent.com/assets/41702/6454982/dd0b4a2e-c10b-11e4-8c6d-1212e28bfe8f.png)

simianhacker commented:
Rolling over the legend only highlights and fades the lines not the dots

![bubble-legend-rollover](https://cloud.githubusercontent.com/assets/41702/6455046/8afc6d7a-c10c-11e4-9870-c2f1fed51ae8.gif)

But it does for the lines (It probably never did the dots but it sure would be nice if it did :D)

![bubble-line-rollover](https://cloud.githubusercontent.com/assets/41702/6455053/a0b8f2be-c10c-11e4-98f6-3054f94f3bd2.gif)

jthomassie commented:
PR #3227 adds highlighting and unhighlighting of the lines, circles and pie
slices on hover and mouseout of the legend (for line, area, column and pie)

On Mon, Mar 2, 2015 at 7:48 PM, Chris Cowan <notifications@github.com>
wrote:

> Rolling over the legend only highlights and fades the lines not the dots
>
> [image: bubble-legend-rollover]
> <https://cloud.githubusercontent.com/assets/41702/6455046/8afc6d7a-c10c-11e4-9870-c2f1fed51ae8.gif>
>
> But it does for the lines (It probably never did the dots but it sure
> would be nice if it did :D)
>
> [image: bubble-line-rollover]
> <https://cloud.githubusercontent.com/assets/41702/6455053/a0b8f2be-c10c-11e4-98f6-3054f94f3bd2.gif>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elasticsearch/kibana/pull/3154#issuecomment-76871146>.
>



-- 

Juan Thomassie
Elasticsearch
m) 703.470.5793
juan.thomassie@elasticsearch.com

panda01 commented:
@simianhacker removed the standard deviation and percentiles metric aggs.

panda01 commented:
@simianhacker there is no min and ma now for changing the dot size, should I add one this PR, or leave that for an enhancement.

rashidkpc commented on src/kibana/components/vislib/visualizations/line_chart.js:
```diff
@@ -73,6 +73,17 @@ define(function (require) {
+        max: _.max(radii)
+      };
+
+      var radiusStep = ((radii.max - radii.min) || (radii.max * 100)) / 20;
```
You should just need to vary the `20` on this line to change the ratio between the min and max dots. A bigger number here should result in a bigger difference between the smallest dots and the biggest ones

panda01 commented:
@simianhacker @rashidkpc Addressed your concerns about the adjustable ratio. 

Added a number input and squared whatever is put in there.

panda01 commented:
Also updated PR details to remark all of the recent changes.

rashidkpc commented:
I'd like to see the dot size ratio adjustment be a slider instead of an input. Also I wonder if it should only be visible if you've actually selected a dot size metric.

I wonder if it might be a good idea to add a bit of alpha transparency to dots when using the dot size metric?

![image](https://cloud.githubusercontent.com/assets/1250387/6514620/0f86e922-c341-11e4-8221-cb8caba04d8d.png)

Also, when the dots get bigger their hover area does not. Only the middle of that circle below is hoverable for tooltip purposes. 

![image](https://cloud.githubusercontent.com/assets/1250387/6514598/e5ebe914-c340-11e4-88ee-63903e6d0900.png)

simianhacker commented:
@rashidkpc I like the look of the transparent dots!

panda01 commented:
@rashidkpc @simianhacker 

Added a hint to the Point radius Label.
Made the point radius thing a slider.
Moved the point radius to inside of the Metric agg as per @spalger's suggestion.
Fixed the hover state with the bigger circles.
I added transparent points, I agree, it does look much cooler!

lukasolson commented:
This is looking really good! The only thing I can see now is that this
![screen shot 2015-03-06 at 11 56 45 am](https://cloud.githubusercontent.com/assets/1178348/6531702/43f9eb16-c3f8-11e4-9883-26c29c59a48f.png)
should probably just use the `<kbn-info>` tag, like it does here: 
![screen shot 2015-03-06 at 11 57 03 am](https://cloud.githubusercontent.com/assets/1178348/6531711/515d93d4-c3f8-11e4-93e4-a43478c69b43.png)

panda01 commented:
@lukasolson Addressed your concerns. 

This thing will be pretty cool when we show it to people.

spalger commented on src/kibana/plugins/visualize/editor/agg.html:
```diff
@@ -54,7 +54,7 @@
       <!-- remove button -->
       <button
-        ng-if="stats.count > stats.min"
+        ng-if="canRemove(agg)"
```
was there a bug here?

spalger replied:
If `canRemove()` is doing something new I cannot tell what it is.

panda01 replied:
`stats.count` is the count of all the aggs, not of all of the aggs of that particular type.
`canRemove()` essentially is just a reduce to make sure there's always the minimum of a particular agg.

panda01 replied:
I will add tests around it.

spalger replied:
We talked this out and `canRemove()` makes sense

rashidkpc commented:
Functionally I'm +1000 on this. Would like @spalger input on the canRemove thing. If we can get rid of it then sweet, otherwise I'd like tests around the functionality.

spalger commented on src/kibana/plugins/vis_types/controls/radius_ratio_option.html:
```diff
@@ -0,0 +1,16 @@
+      </kbn-info>
+    </label>
+  </span>
+  <div class="hintbox" ng-show="showDotRatioHint">
```
This should probably be removed.

lukasolson commented:
LGTM. Assigning to @spalger to continue looking at.

spalger commented on src/kibana/plugins/vis_types/controls/radius_ratio_option.html:
```diff
@@ -0,0 +1,16 @@
+      Dot Size Ratio:
+      <kbn-info
+        placement="right"
+        class="text-info"
```
If we are going to change the color of one kbn-info element we should probably change it for all.

spalger commented on src/kibana/components/vislib/visualizations/line_chart.js:
```diff
@@ -71,10 +71,20 @@ define(function (require) {
+        .flatten()
+        .value();
+
+      radii = {
```
Rather than redefining `radii` here, you could use `_.transform()` at the end of your chain. Especially since the two different versions of radii are of different types.

spalger commented on src/kibana/plugins/vis_types/controls/radius_ratio_option.html:
```diff
@@ -0,0 +1,16 @@
+<div>
+  <span class="hintbox_label" ng-click="showDotRatioHint = !showDotRatioHint">
```
Confused by the use of the `hintbox` classes in this context. There is a css class for `hintbox-label` (not `hintbox_label`) but it doesn't seem to really relate to the dot size ratio at all.

spalger commented on src/kibana/components/vislib/visualizations/line_chart.js:
```diff
@@ -71,10 +71,18 @@ define(function (require) {
+      var radii = _(data)
+        .map(function (series) { return _.map(series, function (point) { return point._input.z; }); })
+        .flatten()
+        .reduce(function (result, val) {
```
:+1:

spalger commented:
LGTM. @panda01 are you planning to add more tests here? You mentioned something, but I'm ready to merge now.

panda01 commented:
I'm adding tests around the canRemove function.

panda01 commented:
Addressed your concerns @spalger.

panda01 commented:
Added making the points opaque only when you don't draw line between points.

spalger commented:
Good work @panda01! Excited to finally have this merged.
